Light mode for Code Hike (#9)
sdorra/sdorra.dev: content/posts/2023-01-12-ch-dark-mode/index.mdx

---
title: Dark mode with Shiki and Code Hike
summary: How to implement beautiful syntax highlighting including dark mode with Shiki and Code Hike using css variables
date: 2023-01-12
image: https://images.unsplash.com/photo-1502725205173-5e6e47cdf175
tags:
  - shiki
  - codehike
  - tailwindcss
  - react
---

<Notification type="warning">
  This article refers to versions of [Code Hike](https://codehike.org/) before
  [0.8.0](https://github.com/code-hike/codehike/releases/tag/v0.8.0). Since version
  [0.8.0](https://github.com/code-hike/codehike/releases/tag/v0.8.0), Code Hike uses
  [lighter](https://github.com/code-hike/lighter) instead of [Shiki](https://shiki.matsu.io/). The approach presented in
  this article uses [shiki's color replacement method](https://github.com/shikijs/shiki/pull/314), which no longer works
  with [lighter](https://github.com/code-hike/lighter).
</Notification>

> I'm using [Code Hike](https://codehike.org/) for the code snippets in my blog and i love this library.

Reviewer: ```suggestion
I'm using [Code Hike](https://codehike.org/) for the code snippets in my blog and I love this library.
```

It has many nice features, like [tabs](https://codehike.org/demo/filenames) or [focus](https://codehike.org/demo/code).

> But it is laking support for switching themes at runtime, which i want to have for my dark mode implementation.

Reviewer: ```suggestion
But it is lacking support for switching themes at runtime, which I want to use for my dark mode implementation.
```

  --shiki-token-punctuation: #c9d1d9;
  --shiki-token-link: #000012;
}
```

**Bonus**: You can use colors from [Tailwind CSS](https://tailwindcss.com/docs/customizing-colors) by using the [theme function](https://tailwindcss.com/docs/functions-and-directives#theme):

```postcss
:root {
  --shiki-color-text: theme("colors.zinc.700");
  --shiki-color-background: theme("colors.white");
  --shiki-token-constant: theme("colors.emerald.700");
  --shiki-token-string: theme("colors.emerald.700");
  --shiki-token-comment: theme("colors.zinc.500");
  --shiki-token-keyword: theme("colors.cyan.700");
  --shiki-token-parameter: theme("colors.pink.700");
  --shiki-token-function: theme("colors.violet.700");
  --shiki-token-string-expression: theme("colors.emerald.700");
  --shiki-token-punctuation: theme("colors.zinc.800");
  --shiki-token-link: theme("colors.zinc.700");
}
```

## Code Hike

Ok, now we have everything together let's test it.
Most of the colors are now loaded from our variables, but not all.

![background and text color not loaded](./assets/01-bg-and-text-not-loaded.png)

The background and the default text color are always very dark.
Why is that?

> After some time of searching in the source code of [Code Hike](https://github.com/code-hike/codehike), i found the answer.

Reviewer: ```suggestion
After some time of searching in the source code of [Code Hike](https://github.com/code-hike/codehike), I found the answer.
```

  "tab.inactiveBackground": "var(--ch-tab-inactive-bg)",
  "icon.foreground": "var(--ch-icon-text)",

  "tab.border": "var(--ch-tab-border)",
  "tab.activeBorder": "var(--ch-tab-active-border)",
},
```

```postcss styles.css
:root {
  --ch-icon-text: theme("colors.zinc.700");
  --ch-tabs-bg: theme("colors.zinc.200");
  --ch-tab-border: theme("colors.zinc.300");
  --ch-tab-active-border: theme("colors.white");
  --ch-tab-active-bg: theme("colors.white");
  --ch-tab-active-color: theme("colors.zinc.700");
  --ch-tab-inactive-color: theme("colors.zinc.500");
  --ch-tab-inactive-bg: theme("colors.zinc.200");
}
```

</CH.Code>

But the color definition for `tab.activeForeground` caused problems.
[Code Hike](https://codehike.org/) tries to convert the color from hex to rgba in order to add opacity to the color ([transparent](https://github.com/code-hike/codehike/blob/next/packages/mdx/src/utils/color.ts#L68)).
But our color definition (`var(--ch-tab-active-color`) is not a hex color code and the function throws an error.
With a bit of css we are able to workaround the problem:

```css
div.ch-editor-tab-active {
  color: var(--ch-tab-active-color);
}
```

> That's it we can now use CSS variables for the style of our syntax highlighting

Reviewer: ```suggestion
That's it, we can now use CSS variables for the style of our syntax highlighting
```

and we can use different themes for light and dark mode.

<Notification type="warning">
  I've tested not all feature of Code Hike. It is very likely that there will still be problems at one point or another.
</Notification>

## Existing Themes

> That is all nice, but how can we use existing VS Code Themes with this approach (Shiki and Code Hike using VS Code Themes).

Reviewer: ```suggestion
That is all nice, but how can we use existing VS Code themes with this approach (Shiki and Code Hike are using VS Code themes).
```
